department-of-veterans-affairs/LEAF
# LEAF 4336 Add Indicator for Special Access Restrictions in the Form Editor

This adds an indicator in the Form Editor if Special Access Restrictions (SAR) are enabled within fields.

Additionally:
- Removes the GET parameter `childkeys`. This doesn't seem to be used anywhere else, and doesn't seem to have any impact when removed.
- Adds the GET parameter `context`. When `context=formEditor`, `./api/form/[FORM ID]` will include an additional `isMaskable` property.
- Added automated test
- Minor verbiage changes
- Removed a couple instances of jQuery's `cache: false` (don't need it because cache control is configured correctly)

### Prerequisite
- The compiled Vue code may conflict with other PRs. After merging, `npm run build` should be executed to recompile it.

### Potential Impact
- The removal of `childkeys=nonnumeric` parameter impacts the ability to re-sort fields. A client-side implementation corrects this. Thanks @aerinkayne for explaining
- Some key logic within Form.php buildFormTree() was modified, intending to simplify it. Need to check record output to ensure it didn't cause any other impact.
- A database query has been added. Performance impact has been mitigated by gating it behind the new GET parameter.

### Testing
- [x] Passes existing API tests
- [ ] After adding SAR via Form Editor -> Open Form -> Edit Field -> Advanced Attributes -> Add group, a lock symbol appears for the field in the form's overview
- [ ] The contents of records are the same as before
- [ ] It's possible to change the order of fields within a form
- [ ] Forms containing Stapled forms sort and preview correctly, e.g.:
  1. Complex form with 2 questions contains a Stapled form with 2 questions
  2. Open the Complex form and click into the Stapled form
  3. Change the sort order of the Stapled form
  4. Click back into the Complex form, and Click Preview
  5. The sort order of the Stapled form in Preview mode should look the same in Edit mode
